File: services/api/src/services/social/GroupSessionService.ts
/**
 * Group Session Service
 * Manages group coaching session lifecycle
 */

import { v4 as uuidv4 } from 'uuid';
import {
  GroupSession,
  GroupSessionAttributes,
  SessionStatus,
  SessionType,
  RecurrencePattern,
} from '../../models/social/GroupSession';
import {
  GroupSessionParticipant,
  ParticipantStatus,
  PaymentStatus,
} from '../../models/social/GroupSessionParticipant';

// ==================== Types ====================

export interface CreateSessionInput {
  coachId: string;
  title: string;
  description: string;
  sessionType: SessionType;
  category: string;
  tags?: string[];
  scheduledAt: Date;
  durationMinutes: number;
  timezone?: string;
  maxParticipants?: number;
  minParticipants?: number;
  isFree?: boolean;
  price?: number;
  currency?: string;
  earlyBirdPrice?: number;
  earlyBirdDeadline?: Date;
  coverImageUrl?: string;
  prerequisites?: string;
  learningObjectives?: string[];
  recordingEnabled?: boolean;
  chatEnabled?: boolean;
  pollsEnabled?: boolean;
  qnaEnabled?: boolean;
  recurrencePattern?: RecurrencePattern;
  recurrenceEndDate?: Date;
}

export interface UpdateSessionInput {
  title?: string;
  description?: string;
  scheduledAt?: Date;
  durationMinutes?: number;
  maxParticipants?: number;
  price?: number;
  coverImageUrl?: string;
  prerequisites?: string;
  learningObjectives?: string[];
  materialsUrl?: string;
  chatEnabled?: boolean;
  pollsEnabled?: boolean;
  qnaEnabled?: boolean;
}

export interface SessionFilters {
  coachId?: string;
  status?: SessionStatus | SessionStatus[];
  sessionType?: SessionType;
  category?: string;
  isFree?: boolean;
  fromDate?: Date;
  toDate?: Date;
  tags?: string[];
  search?: string;
}

export interface PaginationOptions {
  page?: number;
  limit?: number;
  sortBy?: 'scheduledAt' | 'createdAt' | 'price' | 'currentParticipants';
  sortOrder?: 'asc' | 'desc';
}

export interface SessionListResult {
  sessions: GroupSession[];
  total: number;
  page: number;
  limit: number;
  totalPages: number;
}

export interface RegistrationResult {
  success: boolean;
  participant?: GroupSessionParticipant;
  error?: string;
  waitlisted?: boolean;
  requiresPayment?: boolean;
  paymentAmount?: number;
}

// ==================== Service ====================

export class GroupSessionService {
  private sessions: Map<string, GroupSession> = new Map();
  private participants: Map<string, GroupSessionParticipant[]> = new Map();

  // ==================== Session CRUD ====================

  /**
   * Create a new group session
   */
  async createSession(input: CreateSessionInput): Promise<GroupSession> {
    const sessionId = uuidv4();

    const session = {
      id: sessionId,
      coachId: input.coachId,
      title: input.title,
      description: input.description,
      sessionType: input.sessionType,
      category: input.category,
      tags: input.tags ?? [],
      scheduledAt: input.scheduledAt,
      durationMinutes: input.durationMinutes,
      timezone: input.timezone ?? 'UTC',
      recurrencePattern: input.recurrencePattern ?? 'none',
      recurrenceEndDate: input.recurrenceEndDate,
      maxParticipants: input.maxParticipants ?? 20,
      minParticipants: input.minParticipants ?? 1,
      currentParticipants: 0,
      waitlistEnabled: true,
      waitlistCount: 0,
      isFree: input.isFree ?? false,
      price: input.price,
      currency: input.currency ?? 'USD',
      earlyBirdPrice: input.earlyBirdPrice,
      earlyBirdDeadline: input.earlyBirdDeadline,
      status: 'scheduled' as SessionStatus,
      coverImageUrl: input.coverImageUrl,
      prerequisites: input.prerequisites,
      learningObjectives: input.learningObjectives ?? [],
      recordingEnabled: input.recordingEnabled ?? false,
      chatEnabled: input.chatEnabled ?? true,
      pollsEnabled: input.pollsEnabled ?? true,
      qnaEnabled: input.qnaEnabled ?? true,
      reactionsEnabled: true,
      viewCount: 0,
      ratingCount: 0,
      createdAt: new Date(),
      updatedAt: new Date(),
    } as GroupSession;

    this.sessions.set(sessionId, session);
    this.participants.set(sessionId, []);

    // Create recurring sessions if needed
    if (input.recurrencePattern && input.recurrencePattern !== 'none') {
      await this.createRecurringSessions(session);
    }

    return session;
  }

  /**
   * Get session by ID
   */
  async getSession(sessionId: string): Promise<GroupSession | null> {
    return this.sessions.get(sessionId) ?? null;
  }

  /**
   * Update session
   */
  async updateSession(
    sessionId: string,
    input: UpdateSessionInput,
    coachId: string
  ): Promise<GroupSession | null> {
    const session = this.sessions.get(sessionId);
    if (!session || session.coachId !== coachId) return null;

    // Cannot update completed or cancelled sessions
    if (['completed', 'cancelled'].includes(session.status)) {
      throw new Error('Cannot update completed or cancelled sessions');
    }

    Object.assign(session, {
      ...input,
      updatedAt: new Date(),
    });

    this.sessions.set(sessionId, session);
    return session;
  }

  /**
   * List sessions with filters
   */
  async listSessions(
    filters: SessionFilters,
    pagination: PaginationOptions = {}
  ): Promise<SessionListResult> {
    const { page = 1, limit = 20, sortBy = 'scheduledAt', sortOrder = 'asc' } = pagination;

    let sessions = Array.from(this.sessions.values());

    // Apply filters
    if (filters.coachId) {
      sessions = sessions.filter(s => s.coachId === filters.coachId);
    }
    if (filters.status) {
      const statuses = Array.isArray(filters.status) ? filters.status : [filters.status];
      sessions = sessions.filter(s => statuses.includes(s.status));
    }
    if (filters.sessionType) {
      sessions = sessions.filter(s => s.sessionType === filters.sessionType);
    }
    if (filters.category) {
      sessions = sessions.filter(s => s.category === filters.category);
    }
    if (filters.isFree !== undefined) {
      sessions = sessions.filter(s => s.isFree === filters.isFree);
    }
    if (filters.fromDate) {
      sessions = sessions.filter(s => s.scheduledAt >= filters.fromDate!);
    }
    if (filters.toDate) {
      sessions = sessions.filter(s => s.scheduledAt <= filters.toDate!);
    }
    if (filters.tags && filters.tags.length > 0) {
      sessions = sessions.filter(s =>
        filters.tags!.some(tag => s.tags.includes(tag))
      );
    }
    if (filters.search) {
      const search = filters.search.toLowerCase();
      sessions = sessions.filter(
        s =>
          s.title.toLowerCase().includes(search) ||
          s.description.toLowerCase().includes(search)
      );
    }

    // Sort
    sessions.sort((a, b) => {
      let comparison = 0;
      switch (sortBy) {
        case 'scheduledAt':
          comparison = a.scheduledAt.getTime() - b.scheduledAt.getTime();
          break;
        case 'createdAt':
          comparison = a.createdAt.getTime() - b.createdAt.getTime();
          break;
        case 'price':
          comparison = (a.price ?? 0) - (b.price ?? 0);
          break;
        case 'currentParticipants':
          comparison = a.currentParticipants - b.currentParticipants;
          break;
      }
      return sortOrder === 'desc' ? -comparison : comparison;
    });

    const total = sessions.length;
    const totalPages = Math.ceil(total / limit);
    const offset = (page - 1) * limit;
    sessions = sessions.slice(offset, offset + limit);

    return { sessions, total, page, limit, totalPages };
  }

  /**
   * Get upcoming sessions for discovery
   */
  async discoverSessions(
    options: {
      category?: string;
      sessionType?: SessionType;
      isFree?: boolean;
      limit?: number;
    } = {}
  ): Promise<GroupSession[]> {
    const { limit = 10 } = options;

    const result = await this.listSessions(
      {
        status: 'scheduled',
        fromDate: new Date(),
        category: options.category,
        sessionType: options.sessionType,
        isFree: options.isFree,
      },
      { limit, sortBy: 'scheduledAt', sortOrder: 'asc' }
    );

    return result.sessions;
  }

  // ==================== Session Lifecycle ====================

  /**
   * Start a session (go live)
   */
  async startSession(
    sessionId: string,
    coachId: string
  ): Promise<{ success: boolean; meetingUrl?: string; error?: string }> {
    const session = this.sessions.get(sessionId);
    if (!session) return { success: false, error: 'Session not found' };
    if (session.coachId !== coachId) return { success: false, error: 'Unauthorized' };
    if (session.status !== 'scheduled') {
      return { success: false, error: 'Session cannot be started' };
    }

    // Generate meeting credentials
    const meetingId = `gs-${sessionId.slice(0, 8)}`;
    const meetingUrl = `https://meet.upcoach.app/${meetingId}`;
    const meetingPassword = Math.random().toString(36).slice(2, 10);

    session.status = 'live';
    session.meetingId = meetingId;
    session.meetingUrl = meetingUrl;
    session.meetingPassword = meetingPassword;
    session.updatedAt = new Date();

    this.sessions.set(sessionId, session);

    // Notify participants
    await this.notifyParticipants(sessionId, 'session_started', {
      meetingUrl,
      meetingPassword,
    });

    return { success: true, meetingUrl };
  }

  /**
   * End a session
   */
  async endSession(
    sessionId: string,
    coachId: string
  ): Promise<{ success: boolean; error?: string }> {
    const session = this.sessions.get(sessionId);
    if (!session) return { success: false, error: 'Session not found' };
    if (session.coachId !== coachId) return { success: false, error: 'Unauthorized' };
    if (session.status !== 'live') {
      return { success: false, error: 'Session is not live' };
    }

    session.status = 'completed';
    session.updatedAt = new Date();

    this.sessions.set(sessionId, session);

    // Update participant attendance
    await this.finalizeAttendance(sessionId);

    // Notify participants
    await this.notifyParticipants(sessionId, 'session_ended', {});

    return { success: true };
  }

  /**
   * Cancel a session
   */
  async cancelSession(
    sessionId: string,
    coachId: string,
    reason?: string
  ): Promise<{ success: boolean; refundCount?: number; error?: string }> {
    const session = this.sessions.get(sessionId);
    if (!session) return { success: false, error: 'Session not found' };
    if (session.coachId !== coachId) return { success: false, error: 'Unauthorized' };
    if (['completed', 'cancelled'].includes(session.status)) {
      return { success: false, error: 'Session cannot be cancelled' };
    }

    session.status = 'cancelled';
    session.updatedAt = new Date();

    this.sessions.set(sessionId, session);

    // Process refunds
    const refundCount = await this.processRefunds(sessionId);

    // Notify participants
    await this.notifyParticipants(sessionId, 'session_cancelled', { reason });

    return { success: true, refundCount };
  }

  // ==================== Registration ====================

  /**
   * Register for a session
   */
  async registerParticipant(
    sessionId: string,
    userId: string
  ): Promise<RegistrationResult> {
    const session = this.sessions.get(sessionId);
    if (!session) return { success: false, error: 'Session not found' };
    if (session.status !== 'scheduled') {
      return { success: false, error: 'Registration is closed' };
    }

    // Check if already registered
    const participants = this.participants.get(sessionId) ?? [];
    const existing = participants.find(p => p.userId === userId);
    if (existing) {
      return { success: false, error: 'Already registered' };
    }

    // Check capacity
    const isFull = session.currentParticipants >= session.maxParticipants;
    const status: ParticipantStatus = isFull && session.waitlistEnabled
      ? 'waitlisted'
      : 'registered';

    // Determine payment status
    const requiresPayment = !session.isFree;
    const paymentStatus: PaymentStatus = session.isFree ? 'not_required' : 'pending';

    // Calculate price
    let paymentAmount = 0;
    if (requiresPayment) {
      if (
        session.earlyBirdPrice &&
        session.earlyBirdDeadline &&
        new Date() < session.earlyBirdDeadline
      ) {
        paymentAmount = session.earlyBirdPrice;
      } else {
        paymentAmount = session.price ?? 0;
      }
    }

    const participant = {
      id: uuidv4(),
      sessionId,
      userId,
      status,
      role: 'participant',
      paymentStatus,
      currency: session.currency,
      registeredAt: new Date(),
      waitlistPosition: status === 'waitlisted' ? session.waitlistCount + 1 : undefined,
      attendanceMinutes: 0,
      attendancePercentage: 0,
      messageCount: 0,
      reactionCount: 0,
      pollsAnswered: 0,
      questionsAsked: 0,
      handRaiseCount: 0,
      upvoterIds: [],
      createdAt: new Date(),
      updatedAt: new Date(),
    } as GroupSessionParticipant;

    participants.push(participant);
    this.participants.set(sessionId, participants);

    // Update session counts
    if (status === 'registered') {
      session.currentParticipants++;
    } else {
      session.waitlistCount++;
    }
    this.sessions.set(sessionId, session);

    return {
      success: true,
      participant,
      waitlisted: status === 'waitlisted',
      requiresPayment,
      paymentAmount,
    };
  }

  /**
   * Cancel registration
   */
  async cancelRegistration(
    sessionId: string,
    userId: string
  ): Promise<{ success: boolean; refunded?: boolean; error?: string }> {
    const session = this.sessions.get(sessionId);
    if (!session) return { success: false, error: 'Session not found' };

    const participants = this.participants.get(sessionId) ?? [];
    const participantIndex = participants.findIndex(p => p.userId === userId);
    if (participantIndex === -1) {
      return { success: false, error: 'Not registered' };
    }

    const participant = participants[participantIndex];
    const wasRegistered = participant.status === 'registered';
    const wasPaid = participant.paymentStatus === 'completed';

    // Update participant status
    participant.status = 'cancelled';
    participant.updatedAt = new Date();

    // Process refund if paid
    let refunded = false;
    if (wasPaid) {
      participant.paymentStatus = 'refunded';
      participant.refundedAt = new Date();
      participant.refundAmount = participant.amountPaid;
      refunded = true;
    }

    participants[participantIndex] = participant;
    this.participants.set(sessionId, participants);

    // Update session counts
    if (wasRegistered) {
      session.currentParticipants--;

      // Promote from waitlist
      await this.promoteFromWaitlist(sessionId);
    } else if (participant.status === 'waitlisted') {
      session.waitlistCount--;
    }
    this.sessions.set(sessionId, session);

    return { success: true, refunded };
  }

  /**
   * Get session participants
   */
  async getParticipants(
    sessionId: string,
    options: {
      status?: ParticipantStatus;
      includeWaitlist?: boolean;
    } = {}
  ): Promise<GroupSessionParticipant[]> {
    let participants = this.participants.get(sessionId) ?? [];

    if (options.status) {
      participants = participants.filter(p => p.status === options.status);
    }

    if (!options.includeWaitlist) {
      participants = participants.filter(p => p.status !== 'waitlisted');
    }

    return participants;
  }

  /**
   * Check if user is registered
   */
  async isRegistered(sessionId: string, userId: string): Promise<boolean> {
    const participants = this.participants.get(sessionId) ?? [];
    return participants.some(
      p => p.userId === userId && ['registered', 'confirmed', 'attended'].includes(p.status)
    );
  }

  // ==================== Attendance ====================

  /**
   * Record participant joining
   */
  async recordJoin(sessionId: string, userId: string): Promise<void> {
    const participants = this.participants.get(sessionId) ?? [];
    const participant = participants.find(p => p.userId === userId);
    if (!participant) return;

    if (!participant.joinedAt) {
      participant.joinedAt = new Date();
      participant.status = 'attended';
    }

    this.participants.set(sessionId, participants);
  }

  /**
   * Record participant leaving
   */
  async recordLeave(sessionId: string, userId: string): Promise<void> {
    const participants = this.participants.get(sessionId) ?? [];
    const participant = participants.find(p => p.userId === userId);
    if (!participant || !participant.joinedAt) return;

    participant.leftAt = new Date();

    // Calculate attendance
    const attendanceMs = participant.leftAt.getTime() - participant.joinedAt.getTime();
    participant.attendanceMinutes += Math.floor(attendanceMs / 60000);

    const session = this.sessions.get(sessionId);
    if (session) {
      participant.attendancePercentage = Math.min(
        100,
        (participant.attendanceMinutes / session.durationMinutes) * 100
      );
    }

    this.participants.set(sessionId, participants);
  }

  // ==================== Ratings ====================

  /**
   * Submit session rating
   */
  async submitRating(
    sessionId: string,
    userId: string,
    rating: number,
    feedback?: string
  ): Promise<{ success: boolean; error?: string }> {
    if (rating < 1 || rating > 5) {
      return { success: false, error: 'Rating must be between 1 and 5' };
    }

    const session = this.sessions.get(sessionId);
    if (!session) return { success: false, error: 'Session not found' };

    const participants = this.participants.get(sessionId) ?? [];
    const participant = participants.find(p => p.userId === userId);
    if (!participant) {
      return { success: false, error: 'Not a participant' };
    }
    if (participant.rating) {
      return { success: false, error: 'Already rated' };
    }

    participant.rating = rating;
    participant.feedback = feedback;
    participant.feedbackSubmittedAt = new Date();

    this.participants.set(sessionId, participants);

    // Update session average rating
    const allRatings = participants
      .filter(p => p.rating !== undefined)
      .map(p => p.rating!);

    session.ratingCount = allRatings.length;
    session.averageRating =
      allRatings.reduce((sum, r) => sum + r, 0) / allRatings.length;

    this.sessions.set(sessionId, session);

    return { success: true };
  }

  // ==================== Private Helpers ====================

  private async createRecurringSessions(template: GroupSession): Promise<void> {
    if (!template.recurrenceEndDate) return;

    const intervals: Record<string, number> = {
      daily: 1,
      weekly: 7,
      biweekly: 14,
      monthly: 30,
    };

    const interval = intervals[template.recurrencePattern] ?? 0;
    if (interval === 0) return;

    let nextDate = new Date(template.scheduledAt);
    nextDate.setDate(nextDate.getDate() + interval);

    while (nextDate <= template.recurrenceEndDate) {
      await this.createSession({
        ...template,
        scheduledAt: new Date(nextDate),
        recurrencePattern: 'none',
      });

      nextDate.setDate(nextDate.getDate() + interval);
    }
  }

  private async promoteFromWaitlist(sessionId: string): Promise<void> {
    const session = this.sessions.get(sessionId);
    if (!session || session.currentParticipants >= session.maxParticipants) return;

    const participants = this.participants.get(sessionId) ?? [];
    const waitlisted = participants
      .filter(p => p.status === 'waitlisted')
      .sort((a, b) => (a.waitlistPosition ?? 0) - (b.waitlistPosition ?? 0));

    if (waitlisted.length === 0) return;

    const toPromote = waitlisted[0];
    toPromote.status = 'registered';
    toPromote.promotedFromWaitlistAt = new Date();
    toPromote.waitlistPosition = undefined;

    session.currentParticipants++;
    session.waitlistCount--;

    this.sessions.set(sessionId, session);
    this.participants.set(sessionId, participants);

    // Notify promoted participant
    await this.notifyParticipant(toPromote.userId, 'promoted_from_waitlist', {
      sessionId,
      sessionTitle: session.title,
    });
  }

  private async finalizeAttendance(sessionId: string): Promise<void> {
    const session = this.sessions.get(sessionId);
    if (!session) return;

    const participants = this.participants.get(sessionId) ?? [];

    for (const participant of participants) {
      if (participant.joinedAt && !participant.leftAt) {
        // Still in session, record leave now
        await this.recordLeave(sessionId, participant.userId);
      }

      if (!participant.joinedAt && participant.status === 'registered') {
        participant.status = 'no_show';
      }
    }

    this.participants.set(sessionId, participants);
  }

  private async processRefunds(sessionId: string): Promise<number> {
    const participants = this.participants.get(sessionId) ?? [];
    let refundCount = 0;

    for (const participant of participants) {
      if (participant.paymentStatus === 'completed') {
        participant.paymentStatus = 'refunded';
        participant.refundedAt = new Date();
        participant.refundAmount = participant.amountPaid;
        refundCount++;
      }
      participant.status = 'cancelled';
    }

    this.participants.set(sessionId, participants);
    return refundCount;
  }

  private async notifyParticipants(
    sessionId: string,
    eventType: string,
    data: Record<string, any>
  ): Promise<void> {
    const participants = this.participants.get(sessionId) ?? [];
    for (const participant of participants) {
      if (['registered', 'confirmed'].includes(participant.status)) {
        await this.notifyParticipant(participant.userId, eventType, data);
      }
    }
  }

  private async notifyParticipant(
    userId: string,
    eventType: string,
    data: Record<string, any>
  ): Promise<void> {
    // In production, integrate with notification service
    console.log(`[GroupSessionService] Notify ${userId}: ${eventType}`, data);
  }
}

// ==================== Factory ====================

export function createGroupSessionService(): GroupSessionService {
  return new GroupSessionService();
}

export default GroupSessionService;

File: apps/mobile/lib/features/group_coaching/services/group_session_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/group_session_models.dart';

/// Group session service for API communication
class GroupSessionService {
  final String baseUrl;
  final String Function() getToken;

  GroupSessionService({
    required this.baseUrl,
    required this.getToken,
  });

  // ==================== Session Discovery ====================

  /// Get active/upcoming sessions for discovery
  Future<List<GroupSession>> discoverSessions({
    String? category,
    SessionType? sessionType,
    bool? isFree,
    int limit = 10,
  }) async {
    final queryParams = <String, String>{
      'limit': limit.toString(),
    };
    if (category != null) queryParams['category'] = category;
    if (sessionType != null) queryParams['sessionType'] = sessionType.name;
    if (isFree != null) queryParams['isFree'] = isFree.toString();

    final uri = Uri.parse('$baseUrl/group-sessions/discover')
        .replace(queryParameters: queryParams);

    final response = await http.get(
      uri,
      headers: {'Authorization': 'Bearer ${getToken()}'},
    );

    if (response.statusCode == 200) {
      final List<dynamic> data = json.decode(response.body);
      return data.map((json) => GroupSession.fromJson(json)).toList();
    } else {
      throw Exception('Failed to load sessions: ${response.body}');
    }
  }

  /// Get session by ID
  Future<GroupSession> getSession(String sessionId) async {
    final uri = Uri.parse('$baseUrl/group-sessions/$sessionId');

    final response = await http.get(
      uri,
      headers: {'Authorization': 'Bearer ${getToken()}'},
    );

    if (response.statusCode == 200) {
      return GroupSession.fromJson(json.decode(response.body));
    } else {
      throw Exception('Failed to load session: ${response.body}');
    }
  }

  /// List sessions with filters
  Future<List<GroupSession>> listSessions({
    String? coachId,
    SessionStatus? status,
    SessionType? sessionType,
    String? category,
    bool? isFree,
    DateTime? fromDate,
    DateTime? toDate,
    String? search,
    int page = 1,
    int limit = 20,
  }) async {
    final queryParams = <String, String>{
      'page': page.toString(),
      'limit': limit.toString(),
    };

    if (coachId != null) queryParams['coachId'] = coachId;
    if (status != null) queryParams['status'] = status.name;
    if (sessionType != null) queryParams['sessionType'] = sessionType.name;
    if (category != null) queryParams['category'] = category;
    if (isFree != null) queryParams['isFree'] = isFree.toString();
    if (fromDate != null) queryParams['fromDate'] = fromDate.toIso8601String();
    if (toDate != null) queryParams['toDate'] = toDate.toIso8601String();
    if (search != null) queryParams['search'] = search;

    final uri = Uri.parse('$baseUrl/group-sessions')
        .replace(queryParameters: queryParams);

    final response = await http.get(
      uri,
      headers: {'Authorization': 'Bearer ${getToken()}'},
    );

    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      final List<dynamic> sessions = data['sessions'];
      return sessions.map((json) => GroupSession.fromJson(json)).toList();
    } else {
      throw Exception('Failed to load sessions: ${response.body}');
    }
  }

  // ==================== Registration ====================

  /// Register for a session
  Future<RegistrationResult> registerForSession(String sessionId) async {
    final uri = Uri.parse('$baseUrl/group-sessions/$sessionId/register');

    final response = await http.post(
      uri,
      headers: {
        'Authorization': 'Bearer ${getToken()}',
        'Content-Type': 'application/json',
      },
    );

    if (response.statusCode == 200 || response.statusCode == 201) {
      return RegistrationResult.fromJson(json.decode(response.body));
    } else {
      throw Exception('Failed to register: ${response.body}');
    }
  }

  /// Cancel registration
  Future<bool> cancelRegistration(String sessionId) async {
    final uri = Uri.parse('$baseUrl/group-sessions/$sessionId/cancel');

    final response = await http.post(
      uri,
      headers: {
        'Authorization': 'Bearer ${getToken()}',
        'Content-Type': 'application/json',
      },
    );

    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      return data['success'] == true;
    } else {
      throw Exception('Failed to cancel registration: ${response.body}');
    }
  }

  /// Check if user is registered for session
  Future<bool> isRegistered(String sessionId) async {
    final uri = Uri.parse('$baseUrl/group-sessions/$sessionId/is-registered');

    final response = await http.get(
      uri,
      headers: {'Authorization': 'Bearer ${getToken()}'},
    );

    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      return data['isRegistered'] == true;
    } else {
      return false;
    }
  }

  // ==================== Participants ====================

  /// Get session participants
  Future<List<GroupSessionParticipant>> getParticipants(
    String sessionId, {
    ParticipantStatus? status,
    bool includeWaitlist = false,
  }) async {
    final queryParams = <String, String>{
      'includeWaitlist': includeWaitlist.toString(),
    };
    if (status != null) queryParams['status'] = status.name;

    final uri = Uri.parse('$baseUrl/group-sessions/$sessionId/participants')
        .replace(queryParameters: queryParams);

    final response = await http.get(
      uri,
      headers: {'Authorization': 'Bearer ${getToken()}'},
    );

    if (response.statusCode == 200) {
      final List<dynamic> data = json.decode(response.body);
      return data.map((json) => GroupSessionParticipant.fromJson(json)).toList();
    } else {
      throw Exception('Failed to load participants: ${response.body}');
    }
  }

  // ==================== Live Session ====================

  /// Join live session (records attendance)
  Future<void> joinLiveSession(String sessionId) async {
    final uri = Uri.parse('$baseUrl/group-sessions/$sessionId/join');

    final response = await http.post(
      uri,
      headers: {
        'Authorization': 'Bearer ${getToken()}',
        'Content-Type': 'application/json',
      },
    );

    if (response.statusCode != 200) {
      throw Exception('Failed to join session: ${response.body}');
    }
  }

  /// Leave live session
  Future<void> leaveLiveSession(String sessionId) async {
    final uri = Uri.parse('$baseUrl/group-sessions/$sessionId/leave');

    final response = await http.post(
      uri,
      headers: {
        'Authorization': 'Bearer ${getToken()}',
        'Content-Type': 'application/json',
      },
    );

    if (response.statusCode != 200) {
      throw Exception('Failed to leave session: ${response.body}');
    }
  }

  // ==================== Chat ====================

  /// Send chat message
  Future<ChatMessage> sendMessage({
    required String sessionId,
    required String content,
    ChatMessageType type = ChatMessageType.text,
    String? replyToId,
  }) async {
    final uri = Uri.parse('$baseUrl/group-sessions/$sessionId/chat/messages');

    final response = await http.post(
      uri,
      headers: {
        'Authorization': 'Bearer ${getToken()}',
        'Content-Type': 'application/json',
      },
      body: json.encode({
        'content': content,
        'messageType': type.name,
        if (replyToId != null) 'replyToId': replyToId,
      }),
    );

    if (response.statusCode == 200 || response.statusCode == 201) {
      return ChatMessage.fromJson(json.decode(response.body));
    } else {
      throw Exception('Failed to send message: ${response.body}');
    }
  }

  /// Get chat messages
  Future<List<ChatMessage>> getChatMessages(
    String sessionId, {
    int limit = 50,
    String? before,
  }) async {
    final queryParams = <String, String>{
      'limit': limit.toString(),
    };
    if (before != null) queryParams['before'] = before;

    final uri = Uri.parse('$baseUrl/group-sessions/$sessionId/chat/messages')
        .replace(queryParameters: queryParams);

    final response = await http.get(
      uri,
      headers: {'Authorization': 'Bearer ${getToken()}'},
    );

    if (response.statusCode == 200) {
      final List<dynamic> data = json.decode(response.body);
      return data.map((json) => ChatMessage.fromJson(json)).toList();
    } else {
      throw Exception('Failed to load messages: ${response.body}');
    }
  }

  /// React to message
  Future<void> reactToMessage({
    required String sessionId,
    required String messageId,
    required String emoji,
  }) async {
    final uri = Uri.parse(
        '$baseUrl/group-sessions/$sessionId/chat/messages/$messageId/react');

    final response = await http.post(
      uri,
      headers: {
        'Authorization': 'Bearer ${getToken()}',
        'Content-Type': 'application/json',
      },
      body: json.encode({'emoji': emoji}),
    );

    if (response.statusCode != 200) {
      throw Exception('Failed to react to message: ${response.body}');
    }
  }

  /// Upvote message (for Q&A)
  Future<void> upvoteMessage({
    required String sessionId,
    required String messageId,
  }) async {
    final uri = Uri.parse(
        '$baseUrl/group-sessions/$sessionId/chat/messages/$messageId/upvote');

    final response = await http.post(
      uri,
      headers: {
        'Authorization': 'Bearer ${getToken()}',
        'Content-Type': 'application/json',
      },
    );

    if (response.statusCode != 200) {
      throw Exception('Failed to upvote message: ${response.body}');
    }
  }

  // ==================== Polls ====================

  /// Vote on poll
  Future<void> votePoll({
    required String sessionId,
    required String messageId,
    required List<String> optionIds,
  }) async {
    final uri = Uri.parse(
        '$baseUrl/group-sessions/$sessionId/chat/messages/$messageId/vote');

    final response = await http.post(
      uri,
      headers: {
        'Authorization': 'Bearer ${getToken()}',
        'Content-Type': 'application/json',
      },
      body: json.encode({'optionIds': optionIds}),
    );

    if (response.statusCode != 200) {
      throw Exception('Failed to vote: ${response.body}');
    }
  }

  // ==================== Ratings ====================

  /// Submit session rating
  Future<void> submitRating({
    required String sessionId,
    required int rating,
    String? feedback,
  }) async {
    final uri = Uri.parse('$baseUrl/group-sessions/$sessionId/rate');

    final response = await http.post(
      uri,
      headers: {
        'Authorization': 'Bearer ${getToken()}',
        'Content-Type': 'application/json',
      },
      body: json.encode({
        'rating': rating,
        if (feedback != null) 'feedback': feedback,
      }),
    );

    if (response.statusCode != 200) {
      throw Exception('Failed to submit rating: ${response.body}');
    }
  }

  // ==================== User's Sessions ====================

  /// Get user's upcoming sessions
  Future<List<GroupSession>> getMyUpcomingSessions() async {
    final uri = Uri.parse('$baseUrl/group-sessions/my-sessions/upcoming');

    final response = await http.get(
      uri,
      headers: {'Authorization': 'Bearer ${getToken()}'},
    );

    if (response.statusCode == 200) {
      final List<dynamic> data = json.decode(response.body);
      return data.map((json) => GroupSession.fromJson(json)).toList();
    } else {
      throw Exception('Failed to load sessions: ${response.body}');
    }
  }

  /// Get user's session history
  Future<List<GroupSession>> getMySessionHistory() async {
    final uri = Uri.parse('$baseUrl/group-sessions/my-sessions/history');

    final response = await http.get(
      uri,
      headers: {'Authorization': 'Bearer ${getToken()}'},
    );

    if (response.statusCode == 200) {
      final List<dynamic> data = json.decode(response.body);
      return data.map((json) => GroupSession.fromJson(json)).toList();
    } else {
      throw Exception('Failed to load sessions: ${response.body}');
    }
  }
}
